pkg/datapath/maps/map.go

```diff
@@ -34,52 +34,88 @@ import (
 
 var (
 	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "datapath-maps")
+
+	globalSweeper = newMapSweeper(&realEPManager{})
 )
 
-// CollectStaleMapGarbage cleans up stale content in the BPF maps from the
-// datapath.
-func CollectStaleMapGarbage() {
-	if option.Config.DryMode {
-		return
-	}
-	walker := func(path string, _ os.FileInfo, _ error) error {
-		return staleMapWalker(path)
-	}
+// endpointManager checks against its list of the current endpoints to determine
+// whether map paths should be removed, and implements map removal.
+//
+// This interface is provided to abstract epmanager/filesystem access for unit
+// testing.
+type endpointManager interface {
+	endpointExists(endpointID uint16) bool
+	removeDatapathMapping(endpointID uint16) error
+	removeMapPath(path string)
+}
 
-	if err := filepath.Walk(bpf.MapPrefixPath(), walker); err != nil {
-		log.WithError(err).Warn("Error while scanning for stale maps")
+// realEPManager provides an implementatino of endpointManager that is backed
+// by the endpointmanager and policymap packages, and when removeMapPath is
+// invoked, it cleans up paths on the actual filesystem.
+type realEPManager struct{}
+
+func (gw *realEPManager) endpointExists(endpointID uint16) bool {
+	if ep := endpointmanager.LookupCiliumID(endpointID); ep != nil {
+		return true
 	}
+	return false
+}
+
+// removeDatapathMapping unlinks the endpointID from the global policy map, preventing
+// packets that arrive on this node from being forwarded to the endpoint that
+// used to exist with the specified ID.
+func (gw *realEPManager) removeDatapathMapping(endpointID uint16) error {
+	return policymap.RemoveGlobalMapping(uint32(endpointID))
 }
 
-func removeStaleMap(path string) {
+// removeMapPath removes the specified path from the filesystem.
+func (gw *realEPManager) removeMapPath(path string) {
 	if err := os.RemoveAll(path); err != nil {
 		log.WithError(err).WithField(logfields.Path, path).Warn("Error while deleting stale map file")
 	} else {
 		log.WithField(logfields.Path, path).Info("Removed stale bpf map")
 	}
 }
 
-func checkStaleMap(path string, filename string, id string) {
-	if tmp, err := strconv.ParseUint(id, 0, 16); err == nil {
-		if ep := endpointmanager.LookupCiliumID(uint16(tmp)); ep == nil {
-			err2 := policymap.RemoveGlobalMapping(uint32(tmp))
+// mapSweeper is responsible for checking stale map paths on the filesystem
+// and garbage collecting the endpoint if the corresponding endpoint no longer
+// exists.
+type mapSweeper struct {
+	endpointManager
+}
+
+// newMapSweeper creates an object that walks map paths and garbage-collects
+// them.
+func newMapSweeper(g endpointManager) *mapSweeper {
+	return &mapSweeper{
+		endpointManager: g,
+	}
+}
+
+// checkStaleMap uses the garbageWalker implementation to determine for the
+// given path whether it should be deleted, and if so deletes the path.
+func (ms *mapSweeper) checkStaleMap(path string, filename string, endpointID string) {
```

Comment:
While we are here, maybe it makes sense to rename the function to e.g. `delStaleMapIfEPNotExist` to reflect better what it does?